File: neml2_models/src/EffectiveSaturationSecondOrder.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "EffectiveSaturationSecondOrder.h"
#include <neml2/base/Registry.h>

namespace neml2
{
register_NEML2_object(EffectiveSaturationSecondOrder);
OptionSet
EffectiveSaturationSecondOrder::expected_options()
{
  OptionSet options = Model::expected_options();
  options.doc() =
      "Calculate the effective saturation, taking the form of \\f$ S = "
      "\\frac{\\frac{\\phi}{\\phi_\\mathrm{max}} - S_r}{1-S_r} \\f$ where \\f$ \\phi \\f$ is the "
      "volume fraction of the flowing fluid, \\f$ \\phi_\\mathrm{max} \\f$ is the maximum "
      "allowable volume fraction and \\f$ S_r \\f$ is the residual saturation.";

  options.set_parameter<TensorName<Scalar>>("residual_saturation") = "0";
  options.set("residual_saturation").doc() = "Liquid's residual volume fraction";

  options.set_input("fluid_fraction") = VariableName(FORCES, "fluid_fraction");
  options.set("fluid_fraction").doc() = "Volume fraction of the fluid";

  options.set_parameter<TensorName<Scalar>>("max_fraction") = "1";
  options.set("max_fraction").doc() = "Maximum allowable volume fraction of the fluid";

  options.set_output("effective_saturation") = VariableName(STATE, "effective_saturation");
  options.set("effective_saturation").doc() = "Effective saturation";

  options.set_buffer<TensorName<Scalar>>("max_fraction_buffer") = TensorName<Scalar>("0.0001");
  options.set("max_fraction_buffer").doc() = "Buffer to prevent max_fraction to be zero.";

  return options;
}

EffectiveSaturationSecondOrder::EffectiveSaturationSecondOrder(const OptionSet & options)
  : Model(options),
    _Sr(declare_parameter<Scalar>("Sr", "residual_saturation")),
    _phi(declare_input_variable<Scalar>("fluid_fraction")),
    _phimax(declare_parameter<Scalar>("phi_max", "max_fraction", /*allow_nonlinear=*/true)),
    _S(declare_output_variable<Scalar>("effective_saturation")),
    _buf(declare_buffer<Scalar>("phimax_buf", "max_fraction_buffer"))
{
}

void
EffectiveSaturationSecondOrder::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
{
  if (out)
  {
    _S = ( (_phi * _phi) / (_phimax * _phimax + _buf) - _Sr) / (1.0 - _Sr);
  }

  if (dout_din)
  {
    if (_phi.is_dependent())
      _S.d(_phi) = (2.0 * _phi) / ((_phimax * _phimax + _buf) * (1 - _Sr));

    if (const auto * const phimax = nl_param("phi_max"))
      // _S.d(*phimax) = -_phi / (_phimax * _phimax * (1 - _Sr));
      _S.d(*phimax) = - (_phi * _phi) / (1.0 - _Sr) * 2.0 * _phimax / ((_phimax * _phimax + _buf) * (_phimax * _phimax + _buf));
  }
}
}

File: neml2_models/python/EffectiveSaturationSecondOrder.py
from __future__ import annotations

from neml2.factory import register_neml2_object
from neml2.models.chain_rule import ChainRuleDict
from neml2.models.model import Model
from neml2.schema import HitSchema, buffer, input, output, parameter
from neml2.types import Scalar


@register_neml2_object("EffectiveSaturationSecondOrder")
class EffectiveSaturationSecondOrder(Model):
    r"""Second-order effective saturation
    $S = \dfrac{\phi^2 / (\phi_\max^2 + b) - S_r}{1 - S_r}$.
    """

    hit = HitSchema(
        input("fluid_fraction", Scalar, "Volume fraction of the fluid"),
        output("effective_saturation", Scalar, "Effective saturation"),
        parameter("residual_saturation", Scalar, "Liquid's residual volume fraction",
                  attr="Sr", default="0"),
        parameter("max_fraction", Scalar, "Maximum allowable volume fraction of the fluid",
                  attr="phimax", default="1", allow_promotion=True),
        buffer("max_fraction_buffer", Scalar, "Buffer to prevent max_fraction from being zero.",
               attr="buf", default="0.0001"),
    )

    Sr: Scalar
    phimax: Scalar
    buf: Scalar

    def forward(  # type: ignore[override]
        self,
        fluid_fraction: Scalar,
        *promoted_params: Scalar,
        v: ChainRuleDict | None = None,
    ):
        phi = fluid_fraction
        Sr = self._get_param("Sr", promoted_params, Scalar)
        phimax = self._get_param("phimax", promoted_params, Scalar)
        buf = self.buf

        denom = phimax * phimax + buf
        S = ((phi * phi) / denom - Sr) / (1.0 - Sr)
        if v is None:
            return S

        dS_dphi = (2.0 * phi) / (denom * (1.0 - Sr))
        actions = {"fluid_fraction": lambda V, c=dS_dphi: c * V}
        if "phimax" in self._promoted_params:
            dS_dphimax = -(phi * phi) / (1.0 - Sr) * 2.0 * phimax / (denom * denom)
            actions[self._promoted_params["phimax"].input_name] = (
                lambda V, c=dS_dphimax: c * V
            )
        return S, self.apply_chain_rule(v, "effective_saturation", actions, output=S)
